File: algorithms/searching/binary_search.py
"""
binary search works on sorted arrays to divide the array into half until the target is found
Best case: O(n)
Worst case: O(log n)
"""

def binary_search(arr, target, low=0, high=None):
    # calculate high index based on array length
    if high == None:
        high = len(arr) - 1
    
    middle_index = (low + high) // 2

    # check for empty array
    if low > high:
        return -1

    # check if middle equals target
    if target == arr[middle_index]:
        return middle_index
    
    # continue on left side of the array if the target is smaller than middle
    if target < arr[middle_index]:
        return binary_search(arr, target, low, middle_index - 1)

    # continue on the right side of the array if the target is larget than middle
    if target > arr[middle_index]:
        return binary_search(arr, target, middle_index + 1, high)

    # return -1 if element not found
    return -1


array = [5,10,15,20,25,30,35]

print(binary_search(array, 35))

File: c/searching/binary_search.c
#include <stdio.h>

int binary_search(int *arr, int target, int low, int high) {
    int mid = (low + high) / 2;

    if (low > high) {
        return -1;
    }

    // check if mid is the targert
    if (arr[mid] == target) {
        return mid;
    } else if (arr[mid] < target)
    {
        return binary_search(arr, target, mid + 1, high);
    } else if (arr[mid] > target)
    {
        return binary_search(arr, target, low, mid - 1);
    } else {
        return -1;
    }
        
}

int main() {
    int arr[10] = {1,2,3,4,5,6,7,8,9,10};
    int len = sizeof(arr)/sizeof(int);

    int index = binary_search(arr, 7, 0, len-1);
    printf("Target Index is: %d\n", index);
}